<!DOCTYPE html>
<html>
<head>
  <title>snpArcher Track Hub Description</title>
  <style>
    body {
      font-family: Arial, sans-serif;
      margin: 20px;
    }
    h1 {
      font-size: 24px;
      font-weight: bold;
    }
    h2 {
      font-size: 20px;
      font-weight: bold;
      margin-top: 20px;
    }
    h3 {
      font-size: 16px;
      font-weight: bold;
      margin-top: 10px;
      margin-left: 20px;
    }
    p {
      font-size: 16px;
      margin-left: 40px;
    }
  </style>
</head>
<body>
  <h1>snpArcher Track Hub Description</h1>
  
  <h2>Introduction</h2>
  <p>To facilitate downstream data exploration and as an example of the module development components of this work, we
    developed a module to generate UCSC Genome Browser track files to explore population variation data (<a
        href="https://www.biorxiv.org/content/10.1101/2023.06.22.546168v1">see preprint for details</a>).</p>
  
  <h2>Track Descriptions</h2>
  
  <h3>Tajima’s D</h3>
  <p>This track provides windowed estimates of Tajima’s D, a population genetic statistic that measures the departure from neutral evolution in a DNA sequence.</p>
  
  <h3>SNP Density</h3>
  <p>This track displays the density of single nucleotide polymorphisms (SNPs) across the genome, showing regions with high or low levels of genetic variation.</p>
  
  <h3>Pi</h3>
  <p>The Pi track represents the average number of nucleotide differences per site between any two sequences in a population, providing an estimate of genetic diversity.</p>
  
  <h3>Minor Allele Frequency</h3>
  <p>This track shows the frequency of the less common allele at a SNP locus, providing insights into the genetic variation within a population.</p>
  
  <h3>SNP Depth</h3>
  <p>The SNP Depth track displays the number of reads or sequencing depth at each SNP position, indicating the coverage and quality of the variant calls.</p>
  
  <h3>Non Callable Sites</h3>
  <p>The Non Callable Sites track highlights regions in the genome that are considered non-callable, meaning that they have low sequencing coverage or other technical limitations that make it difficult to accurately determine genetic variation in those regions.</p>

</body>
</html>